src/link/MachO.zig

```diff
@@ -282,6 +288,368 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
+    }
+    if (is_dyn_lib) {
+        if ((try rpath_table.fetchPut(full_out_path, {})) == null) {
+            try argv.append("-rpath");
+            try argv.append(full_out_path);
+        }
+    }
+
+    for (self.base.options.lib_dirs) |lib_dir| {
+        try argv.append("-L");
+        try argv.append(lib_dir);
+    }
+
+    // Positional arguments to the linker such as object files.
+    try argv.appendSlice(self.base.options.objects);
+
+    for (comp.c_object_table.items()) |entry| {
+        try argv.append(entry.key.status.success.object_path);
+    }
+    if (module_obj_path) |p| {
+        try argv.append(p);
+    }
+
+    // compiler_rt on darwin is missing some stuff, so we still build it and rely on LinkOnce
+    if (is_exe_or_dyn_lib and !self.base.options.is_compiler_rt_or_libc) {
+        try argv.append(comp.compiler_rt_static_lib.?.full_object_path);
+    }
+
+    // Shared libraries.
+    const system_libs = self.base.options.system_libs.items();
+    try argv.ensureCapacity(argv.items.len + system_libs.len);
+    for (system_libs) |entry| {
+        const link_lib = entry.key;
+        // By this time, we depend on these libs being dynamically linked libraries and not static libraries
+        // (the check for that needs to be earlier), but they could be full paths to .dylib files, in which
+        // case we want to avoid prepending "-l".
+        const ext = Compilation.classifyFileExt(link_lib);
+        const arg = if (ext == .shared_library) link_lib else try std.fmt.allocPrint(arena, "-l{}", .{link_lib});
+        argv.appendAssumeCapacity(arg);
+    }
+
+    // libc++ dep
+    if (!is_obj and self.base.options.link_libcpp) {
+        try argv.append(comp.libcxxabi_static_lib.?.full_object_path);
+        try argv.append(comp.libcxx_static_lib.?.full_object_path);
+    }
+
+    // On Darwin, libSystem has libc in it, but also you have to use it
+    // to make syscalls because the syscall numbers are not documented
+    // and change between versions. So we always link against libSystem.
+    // LLD craps out if you do -lSystem cross compiling, so until that
+    // codebase gets some love from the new maintainers we're left with
+    // this dirty hack.
+    if (self.base.options.is_native_os) {
+        try argv.append("-lSystem");
+    }
+
+    for (self.base.options.framework_dirs) |framework_dir| {
+        try argv.append("-F");
+        try argv.append(framework_dir);
+    }
+    for (self.base.options.frameworks) |framework| {
+        try argv.append("-framework");
+        try argv.append(framework);
+    }
+
+    if (allow_shlib_undefined) {
+        try argv.append("-undefined");
+        try argv.append("dynamic_lookup");
+    }
+    if (self.base.options.bind_global_refs_locally) {
+        try argv.append("-Bsymbolic");
+    }
+
+    if (self.base.options.verbose_link) {
+        Compilation.dump_argv(argv.items);
+    }
+
+    // TODO allocSentinel crashed stage1 so this is working around it.
```

Comment:
#6370 is now fixed, `allocSentinel` should work just fine now

Reply:
Nice!